torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -301,43 +264,98 @@ static bool CheckEqual(const Expr* lhs, const Expr* rhs) {
   return immediateEquals(diff_s.node(), 0);
 }
 
-// Identify the pattern: a[e1] = a[e1] + e2.
-static bool isAtomicAdd(const Store* v, const Expr** atomic_add_value) {
-  ScalarType dtype = v->value()->dtype().scalar_type();
-  if (dtype != ScalarType::Float && dtype != ScalarType::Double) {
-    return false;
-  }
-  const Add* add_v = dynamic_cast<const Add*>(v->value());
-  if (!add_v) {
-    return false;
-  }
-  const Load* load_v = dynamic_cast<const Load*>(add_v->lhs());
-  if (!load_v) {
-    return false;
-  }
-  if (v->base_handle() != load_v->base_handle()) {
-    return false;
-  }
-  if (v->indices().empty() && load_v->indices().empty()) {
-    return false;
-  }
-  bool index_equal = CheckEqual(v->flat_index(), load_v->flat_index());
-  if (index_equal) {
-    *atomic_add_value = add_v->rhs();
+class AtomicAddFuser : public IRMutator {
+ public:
+  AtomicAddFuser(
+      const std::unordered_set<const Var*>& thread_local_bufs,
+      const GPUMetaVarRewriter* metavars)
+      : thread_local_bufs_(thread_local_bufs) {
+    size_t DIMS = 3;
+
+    const std::vector<const Expr*>& block_extents =
```

Comment:
Minor: Style wise, I mildly prefer "std::vector<const Expr*> block_extents = ...", over "const ... &" here. With return-value-optimization, both does the same thing. But the first is easier to tell the scope of the temporary object, while the second relies on a special rule in C++ to extend its lifespan. Both are correct, just some people might trip over why the 2nd form is correct. 

Either way is fine with me, if you are more comfortable with that style.

Reply:
Not particularly bothered either way, I can switch to value.